src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -250,6 +264,42 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
             throw ExceptionUtilities.UnexpectedValue(method);
         }
 
+        private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F, object builderAttributeArgument, Accessibility desiredAccessibility, int desiredArity)
+        {
+            var builderType = builderAttributeArgument as NamedTypeSymbol;
+
+            bool isArityOk;
+            if (desiredArity == 0)
+            {
+                isArityOk = builderType?.IsGenericType == false;
+            }
+            else if (desiredArity == 1)
+            {
+                isArityOk = builderType?.IsUnboundGenericType == true &&
+                            builderType?.ContainingType?.IsGenericType != true;
+            }
+            else
+            {
+                isArityOk = false;
+                Debug.Assert(false, "expected arity 0 or 1");
+            }
+
+
+            if ((object)builderType == null ||
+                 builderType.IsErrorType() == true ||
+                 builderType.SpecialType == SpecialType.System_Void ||
+                 builderType.DeclaredAccessibility != desiredAccessibility ||
```

Comment:
There was a bit of discussion on this. I was taken by this comment: https://github.com/dotnet/roslyn/issues/10902#issuecomment-242844274

I don't think it's "ideal" to allow an internal builder type on a public tasklike. That would mean, as a consumer, I see a type which is tasklike in some respects (overload resolution + type inference), but isn't tasklike in the sense that I can return it from an async method/lambda. But the impact of tasklikeness in type inference and (most of) overload resolution only has effect on async lambdas. It'd also mean that an async method can return this tasklike in one place but not another. It feels like a source of confusion for no gain.

In the second case `internal class C { public class B { } }` the accessibility domain of `B` is exactly the same as if you'd written `internal class B { }`.

The way I see it, this is a marginal area to start with, and there's a slight preference for the two types to have the same accessibility domain. The implementation of checking declared accessibility achieves almost that, with the simplest code. We could try to support your second example of *nested public B* but that wouldn't gain any scenarios, and would come at the expense of more code, so it's not worth it.